services/src/datasets/external/netcdfcf/mod.rs

```diff
@@ -0,0 +1,1109 @@
+
+        Ok(unsafe {
+            let c_mdarray = GDALGroupOpenMDArray(self.c_group, name.as_ptr(), options.as_ptr());
+
+            if c_mdarray.is_null() {
+                return Err(NetCdfCf4DProviderError::GdalMd {
+                    source: Self::_last_null_pointer_err("GDALGroupOpenMDArray"),
+                });
+            }
+
+            MdArray {
+                c_mdarray,
+                _group: self,
+            }
+        })
+    }
+}
+/// TODO: There are potential memory leaks from the C-API
+impl<'g> MdArray<'g> {
+    fn spatial_reference(&self) -> Result<SpatialReferenceOption> {
+        let gdal_spatial_ref = unsafe {
+            let c_gdal_spatial_ref = GDALMDArrayGetSpatialRef(self.c_mdarray);
+
+            gdal::spatial_ref::SpatialRef::from_c_obj(c_gdal_spatial_ref)
+        }
+        .context(error::GdalMd)?;
+
+        SpatialReference::try_from(gdal_spatial_ref)
+            .map(Into::into)
+            .context(error::CannotParseCrs)
+    }
+
+    fn data_type(&self) -> Result<RasterDataType> {
+        let gdal_data_type = unsafe {
+            let c_data_type = GDALMDArrayGetDataType(self.c_mdarray);
+
+            GDALExtendedDataTypeGetNumericDataType(c_data_type)
+        };
+
+        Ok(match gdal_data_type {
+            GDALDataType::GDT_Byte => RasterDataType::U8,
+            GDALDataType::GDT_UInt16 => RasterDataType::U16,
+            GDALDataType::GDT_Int16 => RasterDataType::I16,
+            GDALDataType::GDT_UInt32 => RasterDataType::U32,
+            GDALDataType::GDT_Int32 => RasterDataType::I32,
+            GDALDataType::GDT_Float32 => RasterDataType::F32,
+            GDALDataType::GDT_Float64 => RasterDataType::F64,
+            _ => {
+                return Err(NetCdfCf4DProviderError::UnknownGdalDatatype {
+                    type_number: gdal_data_type,
+                })
+            }
+        })
+    }
+
+    fn no_data_value(&self) -> Option<f64> {
+        let mut has_nodata = 0;
+
+        let no_data_value = unsafe {
+            GDALMDArrayGetNoDataValueAsDouble(self.c_mdarray, &mut has_nodata as *mut c_int)
+        };
+
+        if has_nodata == 0 {
+            None
+        } else {
+            Some(no_data_value)
+        }
+    }
+
+    fn dimensions(&self) -> Result<DimensionSizes> {
+        let mut number_of_dimensions = 0;
+
+        Ok(unsafe {
+            let c_dimensions =
+                GDALMDArrayGetDimensions(self.c_mdarray, &mut number_of_dimensions as *mut usize);
+
+            if number_of_dimensions != 4 {
+                return Err(NetCdfCf4DProviderError::MustBe4DDataset {
+                    number_of_dimensions,
+                });
+            }
+
+            let dimensions = std::slice::from_raw_parts_mut(c_dimensions, number_of_dimensions);
+
+            DimensionSizes {
+                _entity: GDALDimensionGetSize(dimensions[0]) as usize,
+                time: GDALDimensionGetSize(dimensions[1]) as usize,
```

Comment:
is hte order fix? only for our case or all CF netcdfs??

Reply:
CF enforces this order